### PR TITLE
Dashboard Save/Revert Flow

### PR DESCRIPTION
**Pull request by Alex-Aralis**
## Description
Change Dashboards to require a manual save to be updated.  Also allow dashboards to be reverted when in a dirty state.

## Proposed Changes
- Add a save/revert foot to dashboard view when dirty.
- Stop auto saving dashboard on change.

## Screenshots
image

image

### CODE REVIEW

File: packages/dashboards/addon/routes/dashboards/dashboard.js [hunk 1]
```diff
@@ -107,57 +91,55 @@ export default Route.extend({
...
      * @action saveDashboard - saves dashboard updates
      */
     saveDashboard() {
-      run.debounce(this, this._saveDashboardFn, 100);
+      const dashboard = get(this, 'currentDashboard');
+      const widgets = get(this, 'currentDashboard.widgets');
+
+      return dashboard
+        .save()
+        .then(all(widgets.map(async widget => widget.hasDirtyAttributes && widget.save())))
```
Comment (mattgraham1995) [inline packages/dashboards/addon/routes/dashboards/dashboard.js]: What happens if a widget fails to save?

File: packages/dashboards/addon/templates/components/navi-dashboard.hbs [hunk 2]
```diff
@@ -119,6 +119,8 @@
           width=dashboard.presentation.columns
         )
         onChange=(route-action "didUpdateLayout")
+        onDragstop=(route-action "commitStagedLayout")
```
Comment (mattgraham1995) [inline packages/dashboards/addon/templates/components/navi-dashboard.hbs]: `onDragStop`, `onResizeStop`

File: packages/dashboards/addon/templates/components/navi-dashboard.hbs [hunk 3]
```diff
@@ -155,3 +157,10 @@
     </div>
   {{/if}}
 </div>
+
+{{#if dashboard.hasDirtyAttributes}}
+  <div class="navi-dashboard__save-dialogue">
+    <button onclick={{route-action "saveDashboard"}} class="btn btn-secondary navi-dashboard__save-button">Save Changes</button>
```
Comment (mattgraham1995) [inline packages/dashboards/addon/templates/components/navi-dashboard.hbs]: Should save be a primary button?

File: packages/dashboards/addon/routes/dashboards/dashboard.js (hunk 1 above)
Comment (Alex-Aralis) [reply packages/dashboards/addon/routes/dashboards/dashboard.js]: The failure will propagate out to the catch at the end of the chain and give an error message.

If a widget is delete from a dashboard and the dashboard saves, but the widget fails to delete then the dashboard will own a widget that is not referenced in its layout.  The alternative is to have a layout that references a deleted widget.  This seems like a better alternative.

File: packages/dashboards/tests/acceptance/dashboards-test.js [hunk 4]
```diff
@@ -433,4 +450,52 @@ module('Acceptance | Dashboards', function(hooks) {
         'Unauthorized widget loaded unauthorized component'
       );
   });
+
+  test('dashboard save/revert', async function(assert) {
+    assert.expect(10);
+
+    server.patch('/dashboards/1', () => {
+      assert.notok(true, 'Dashboard should not be patched before save.');
```
Comment (mattgraham1995) [inline packages/dashboards/tests/acceptance/dashboards-test.js]: `assert.notOk`

File: packages/dashboards/tests/acceptance/dashboards-test.js [hunk 5]
```diff
@@ -433,4 +450,52 @@ module('Acceptance | Dashboards', function(hooks) {
...
+
+    await click('.editable-label__icon');
+    await fillIn('.editable-label__input', 'EFGH');
+    await blur('.editable-label__input');
+
+    assert.dom('.navi-dashboard__save-button').isVisible('Save appears after change');
+
+    server.patch('/dashboards/1', () => {
+      assert.ok(true, 'Dashboard should be patched after save.');
+      return new Response(500);
```
Comment (mattgraham1995) [inline packages/dashboards/tests/acceptance/dashboards-test.js]: Do we want to return a 500 here?

File: packages/dashboards/tests/dummy/app/services/navi-notifications.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright 2017, Yahoo Holdings Inc.
```
Comment (mattgraham1995) [inline packages/dashboards/tests/dummy/app/services/navi-notifications.js]: 2019

File: packages/dashboards/addon/templates/components/navi-dashboard.hbs (hunk 3 above)
Comment (Alex-Aralis) [reply packages/dashboards/addon/templates/components/navi-dashboard.hbs]: Not according to the mockup.

image

File: packages/dashboards/tests/acceptance/dashboards-test.js (hunk 5 above)
Comment (Alex-Aralis) [reply packages/dashboards/tests/acceptance/dashboards-test.js]: The app freaks out if a 200 returns with no body.

**Pull request merged**